### PR TITLE
Symintify decomps for split and upsample_bilinear; Fix decomp for _softmax_backward_data and native_dropout_backward

### PR DESCRIPTION
**Pull request by SherlockNoMad**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88761

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_decomp/decompositions.py [hunk 1]
```diff
@@ -696,7 +697,7 @@ def _softmax_backward_data(
     grad_input = new_grad_output - output * torch.sum(
         new_grad_output, dim=dim, keepdim=True
     )
-    return _cast_grad_to_input_dtype(grad_output, grad_input, input_dtype)
+    return _cast_grad_to_input_dtype(grad_output, grad_input, input_dtype).contiguous()
```
Comment (ezyang) [inline torch/_decomp/decompositions.py:700]: need test for this fix

File: torch/_decomp/decompositions.py [hunk 2]
```diff
@@ -914,7 +915,13 @@ def check_positive(param, param_name, strict=True):
...
 @pw_cast_for_opmath
 def native_dropout_backward(grad_output: Tensor, mask: Tensor, scale: float):
-    return grad_output * (mask.type_as(grad_output) * scale)
+    # According to the CUDA kernel implementation we should have this test;
+    # but it seems to fail tests!
+    # utils.check(mask.dtype == torch.bool, lambda: f"Mask should be Bool Scalar Type {mask.dtype}")
+    r = (grad_output * (mask.type_as(grad_output) * scale)).clone(
+        memory_format=utils.suggest_memory_format(grad_output)
+    )
+    return r
```
Comment (ezyang) [inline torch/_decomp/decompositions.py:924]: test

File: torch/_decomp/decompositions.py [hunk 3]
```diff
@@ -914,7 +915,13 @@ def check_positive(param, param_name, strict=True):
 @register_decomposition(aten.native_dropout_backward)
 @pw_cast_for_opmath
 def native_dropout_backward(grad_output: Tensor, mask: Tensor, scale: float):
-    return grad_output * (mask.type_as(grad_output) * scale)
+    # According to the CUDA kernel implementation we should have this test;
+    # but it seems to fail tests!
+    # utils.check(mask.dtype == torch.bool, lambda: f"Mask should be Bool Scalar Type {mask.dtype}")
+    r = (grad_output * (mask.type_as(grad_output) * scale)).clone(
+        memory_format=utils.suggest_memory_format(grad_output)
```
Comment (anjali411) [inline torch/_decomp/decompositions.py:922]: this seems like a fix unrelated to symintification right?

File: torch/_decomp/decompositions.py [hunk 4]
```diff
@@ -1830,31 +1838,31 @@ def upsample_bilinear2d_vec(
     n_batch, n_channels, in_h, in_w = input.shape
 
     if output_size is not None:
-        out_h = float(output_size[0])
-        out_w = float(output_size[1])
+        out_h = sym_float(output_size[0])
+        out_w = sym_float(output_size[1])
```
Comment (ezyang) [inline torch/_decomp/decompositions.py:1842]: test

File: torch/_decomp/decompositions.py [hunk 5]
```diff
@@ -1095,8 +1102,9 @@ def split(self: Tensor, split_size: int, dim: int = 0) -> List[Tensor]:
         assert dim_size == 0
         return [self]
     chunks = (dim_size + split_size - 1) // split_size
+    chunks = guard_int(chunks)
     split_sizes = [split_size for i in range(chunks)]
-    split_sizes[chunks - 1] = split_size - (split_size * chunks - dim_size)
+    split_sizes[-1] = split_size - (split_size * chunks - dim_size)
     return torch.split(self, split_sizes, dim)
```
Comment (anjali411) [inline torch/_decomp/decompositions.py:1108]: which tests do we currently run with decomps?

Comment (ezyang) [reply torch/_decomp/decompositions.py:1108]: split xfail was removed with this change, so this seems reasonable to me

File: torch/_decomp/decompositions.py (hunk 2 above)
Comment (SherlockNoMad) [reply torch/_decomp/decompositions.py:923]: added opinfo

File: torch/_decomp/decompositions.py (hunk 5 above)
Comment (SherlockNoMad) [reply torch/_decomp/decompositions.py:1107]: this is tested.

File: torch/_decomp/decompositions.py (hunk 4 above)
Comment (SherlockNoMad) [reply torch/_decomp/decompositions.py:1841]: opinfo has nn.function.upsample_bilinear2d, so this is tested with real int. 
Do you mean we need additional test case for SymInt output_size.

Comment (ezyang) [reply torch/_decomp/decompositions.py:1841]: No, just that I didn't see an xfail removal for this test. So maybe we do need more inputs

Comment (ezyang) [reply torch/_decomp/decompositions.py:1846]: cc @albanD who did something similar for Sizes

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -1159,7 +1174,7 @@ def sample_inputs_zero_(op_info, device, dtype, requires_grad, **kwargs):
     cases = ((), (S, S, S), (S,))
 
     for shape in cases:
-        yield(SampleInput(make_arg(shape)))
+        yield (SampleInput(make_arg(shape)))
```
Comment (ezyang) [inline torch/testing/_internal/common_methods_invocations.py:1177]: fwiw you don't need the parens either

File: torch/testing/_internal/common_methods_invocations.py [hunk 7]
```diff
@@ -406,6 +406,21 @@ def sample_inputs_batch_norm(op_info, device, dtype, requires_grad, **kwargs):
...
+        ((S,), 0),
+        ((S, S), 0),
+        ((S, M, S), -1),
+    ]
+    input_dtypes = [dtype]
+    if dtype == torch.float and device == 'cuda':
+        input_dtypes += [torch.float16]
+
+    for (shape, dim), input_dtype in product(cases, input_dtypes):
+        yield SampleInput(make_arg(shape), make_arg(shape), dim, input_dtype)
```
Comment (ezyang) [inline torch/testing/_internal/common_methods_invocations.py:423]: Are you afraid these samples overlap with the softmax sample inputs?

**Review by ezyang — APPROVED** [review]

File: torch/_decomp/decompositions.py (hunk 4 above)
Comment (albanD) [reply torch/_decomp/decompositions.py:1846]: This was done in https://github.com/pytorch/pytorch/pull/87304

File: torch/_decomp/decompositions.py (hunk 3 above)
Comment (SherlockNoMad) [reply torch/_decomp/decompositions.py:926]: yes, but it's still needed for correct stride.

Comment (ngimel) [reply torch/_decomp/decompositions.py:926]: hm, this looks wrong, dropout propagates strides, so even if grad_output is not channels last, the permutation of input/grad_output would be preserved.

Comment (SherlockNoMad) [reply torch/_decomp/decompositions.py:926]: For native_dropout_backward cuda kernel, the output stride doesn't follow TI's behavior. 
I suspect this is calling cudnn dropout backward. 

```
>>> torch.ops.aten.native_dropout_backward(r.permute(2,3,1,0), m, 0.2).stride()
(8, 4, 2, 1)
>>> torch.ops.aten.native_dropout_backward(r.permute(0,3,1,2), m, 0.2).stride()
(8, 1, 4, 2)
```

Comment (SherlockNoMad) [reply torch/_decomp/decompositions.py:926]: For CPU kernel, it's following TI's behavior.

Comment (SherlockNoMad) [reply torch/_decomp/decompositions.py:926]: CUDA's impl 

```
template <typename mask_t>
Tensor dropout_backward_cuda(const Tensor& grad, const Tensor& mask, double scale){
   Tensor ret = at::empty_like(grad, grad.suggest_memory_format());
   AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, ret.scalar_type(), "masked_scale", [&] {
      using accscalar_t = acc_type<scalar_t, true>;
      masked_scale_kernel<mask_t, scalar_t>(ret, grad, mask, (accscalar_t)scale);
  });
  return ret;
}
```

**Comment by anjali411** [comment]
@SherlockNoMad some jobs on master seem to be failing after this PR due to two unexpected successes in `dynamo/test_dynamic_shapes.py`. Not sure why they weren't caught on the CI because I thought we were running some dynamo tests on CI ... anyway, do you want to land a forward fix?

**Comment by SherlockNoMad** [comment]
@pytorbot rebase

File: torch/_decomp/decompositions.py (hunk 3 above)
Comment (ngimel) [reply torch/_decomp/decompositions.py:929]: Ah I see, usually for cuda dropout `fused_dropout`/`fused_dropout_backward` (that dispatches to `masked_scale`) is used, and that one has a more reasonable stride propagation behavior.

Comment (ngimel) [reply torch/_decomp/decompositions.py:929]: Hm, no, I'm confused, how does this work:
```
In [2]: a=torch.randn(1024, 1024, device="cuda").t()

In [4]: gO=torch.rand_like(out[0])

In [5]: a.requires_grad_(True)
In [6]: out=torch.native_dropout(a, p=0.5, train=True)

In [8]: out[0].backward(gO)

In [9]: a.grad.stride()
Out[9]: (1, 1024)
```
gO here can't have `suggest_memory_format` suggest anything, does permutation happen after the kernel? That's weird.

**Review by ezyang — APPROVED** [review]

**Comment by kit1980** [comment]
@SherlockNoMad This PR has broken [linux-focal-rocm5.2-py3.8-slow / test (slow, 1, 1, linux.rocm.gpu)](https://github.com/pytorch/pytorch/actions/runs/3483501734/jobs/5827293170#logs)

Two test fail: test_nvfuser_correctness__softmax_backward_data_cuda_bfloat16 and test_nvfuser_correctness__softmax_backward_data_cuda_float16

Because the broken workflow is a periodic workflow, it would be better to forward fix this instead of reverting.
